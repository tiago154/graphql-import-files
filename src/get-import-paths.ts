/**
 * Checks for external files imported by the #import annotation in a file and returns the paths if any
 * @param {string} fileContent
 * @returns {string[] | []} The list of file paths with #import notation
 */
export const getImportPaths = (fileContent: string): string[] | [] =>
  fileContent
    .split(/\r?\n/)
    .filter((line) => line.match(/^#import\s/g))
    .map(line => line.match(/'.*'|".*"/g)!.toString().replace(/'|"/g, ''))
