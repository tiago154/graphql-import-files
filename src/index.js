const loadFile = require('./loadFile')
const loadFiles = require('./loadFiles')

module.exports = { loadFile, loadFiles }
