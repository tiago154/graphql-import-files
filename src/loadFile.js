const fs = require('fs')
const path = require('path')

/**
 * Place the path of the file to be imported, considering the root of the project.
 * @example ./schemas/file.graphql
 * @param {string} pathFile
 * @return {string} String of a scheme
 */
module.exports = pathFile =>
  fs.readFileSync(path.join(process.cwd(), pathFile), 'utf8')