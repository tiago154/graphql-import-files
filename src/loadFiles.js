const fs = require('fs')
const path = require('path')
const glob = require('glob')
const { mergeTypeDefs } = require('@graphql-toolkit/schema-merging')


/**
 * Loads multiple graphql files, use .graphql or .gql.
 * @param {string} matchFiles - The glob patterns uses to detect graphql files
 * @return {DocumentNode} DocumentNode
 */
module.exports = matchFiles => {
  const files = glob.sync(matchFiles)

  const schema = files.map(file =>
    fs.readFileSync(path.join(process.cwd(), file), 'utf8')
  )

  return mergeTypeDefs(schema)
}