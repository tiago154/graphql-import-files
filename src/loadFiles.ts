import { readFileSync } from 'fs'
import { join } from 'path'

import glob from 'fast-glob'
import { mergeTypeDefs } from '@graphql-tools/merge'
import type { DocumentNode } from 'graphql'

/**
 * Loads multiple graphql files, use .graphql or .gql.
 * @param {string} matchFiles - The glob patterns uses to detect graphql files
 * @return {object} DocumentNode
 */
export const loadFiles = (matchFiles: string): DocumentNode => {
  const files = glob.sync(matchFiles)

  const schema = files.map(file =>
    readFileSync(join(process.cwd(), file), 'utf8')
  )

  return mergeTypeDefs(schema)
}
