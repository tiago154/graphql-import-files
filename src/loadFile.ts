import { readFileSync } from 'fs'
import { join, dirname } from 'path'

import { mergeTypeDefs } from '@graphql-tools/merge'
import type { DocumentNode } from 'graphql'

import { getImportPaths } from './get-import-paths'

/**
 * Put the path of the file to be imported, considering the root of the project i.e. the directory from where you invoked the node command
 * @example ./schemas/file.graphql
 * @param {string} pathFile
 * @return {DocumentNode} DocumentNode
 */
export const loadFile = (pathFile: string): DocumentNode => {
  const fileContent = readFileSync(join(process.cwd(), pathFile), 'utf8')

  const externalFilePaths = getImportPaths(fileContent)

  const contentExternalFiles = externalFilePaths.map(externalFilePath => readFileSync(
    join(process.cwd(), dirname(pathFile), externalFilePath),
    'utf8'
  ))

  return mergeTypeDefs([
    fileContent,
    ...contentExternalFiles
  ])
}
