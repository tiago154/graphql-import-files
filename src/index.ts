import { loadFile } from './loadFile'
import { loadFiles } from './loadFiles'

export { loadFile, loadFiles }
