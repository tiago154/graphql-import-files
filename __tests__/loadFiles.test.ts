import { loadFiles } from '../src'

describe('loadFiles', () => {
  describe('success', () => {
    it('should valid schema with many file', () => {
      const globPattern = '**/__tests__/mocks/multifiles/**/*.{graphql,gql}'
      const schema = loadFiles(globPattern)

      expect(schema).toMatchSnapshot()
    })

    it('should build the schema even with files with separate queries', () => {
      const globPattern = '**/__tests__/mocks/separate-queries/*.graphql'
      const schema = loadFiles(globPattern)

      expect(schema).toMatchSnapshot()
    })
  })

  describe('failure', () => {
    it('should returns an object with definitions empty when no file is founds', () => {
      const glopPattern = '**/__tests__/mocks/multifiles/**/*.{error,fail}'
      const schema = loadFiles(glopPattern)

      expect(schema).toEqual({ definitions: [], kind: 'Document' })
    })
  })
})
