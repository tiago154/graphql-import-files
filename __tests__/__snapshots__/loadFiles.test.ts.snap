// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadFiles success should build the schema even with files with separate queries 1`] = `
GraphQLSchema {
  "__validationErrors": undefined,
  "_directives": Array [
    "@skip",
    "@include",
    "@deprecated",
    "@specifiedBy",
  ],
  "_implementationsMap": Object {},
  "_mutationType": undefined,
  "_queryType": "Query",
  "_subTypeMap": Object {},
  "_subscriptionType": undefined,
  "_typeMap": Object {
    "Boolean": "Boolean",
    "Int": "Int",
    "Query": "Query",
    "String": "String",
    "User": "User",
    "__Directive": "__Directive",
    "__DirectiveLocation": "__DirectiveLocation",
    "__EnumValue": "__EnumValue",
    "__Field": "__Field",
    "__InputValue": "__InputValue",
    "__Schema": "__Schema",
    "__Type": "__Type",
    "__TypeKind": "__TypeKind",
  },
  "astNode": Object {
    "description": undefined,
    "directives": Array [],
    "kind": "SchemaDefinition",
    "loc": Object {
      "end": 23,
      "start": 0,
    },
    "operationTypes": Array [
      Object {
        "kind": "OperationTypeDefinition",
        "loc": Object {
          "end": 21,
          "start": 9,
        },
        "operation": "query",
        "type": Object {
          "kind": "NamedType",
          "loc": Object {
            "end": 21,
            "start": 16,
          },
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 21,
              "start": 16,
            },
            "value": "Query",
          },
        },
      },
    ],
  },
  "description": undefined,
  "extensionASTNodes": Array [],
  "extensions": undefined,
}
`;

exports[`loadFiles success should valid schema with many file 1`] = `
GraphQLSchema {
  "__validationErrors": undefined,
  "_directives": Array [
    "@skip",
    "@include",
    "@deprecated",
    "@specifiedBy",
  ],
  "_implementationsMap": Object {},
  "_mutationType": undefined,
  "_queryType": "Query",
  "_subTypeMap": Object {},
  "_subscriptionType": undefined,
  "_typeMap": Object {
    "Boolean": "Boolean",
    "Float": "Float",
    "ID": "ID",
    "Int": "Int",
    "Product": "Product",
    "Query": "Query",
    "String": "String",
    "User": "User",
    "__Directive": "__Directive",
    "__DirectiveLocation": "__DirectiveLocation",
    "__EnumValue": "__EnumValue",
    "__Field": "__Field",
    "__InputValue": "__InputValue",
    "__Schema": "__Schema",
    "__Type": "__Type",
    "__TypeKind": "__TypeKind",
  },
  "astNode": Object {
    "description": undefined,
    "directives": Array [],
    "kind": "SchemaDefinition",
    "loc": Object {
      "end": 23,
      "start": 0,
    },
    "operationTypes": Array [
      Object {
        "kind": "OperationTypeDefinition",
        "loc": Object {
          "end": 21,
          "start": 9,
        },
        "operation": "query",
        "type": Object {
          "kind": "NamedType",
          "loc": Object {
            "end": 21,
            "start": 16,
          },
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 21,
              "start": 16,
            },
            "value": "Query",
          },
        },
      },
    ],
  },
  "description": undefined,
  "extensionASTNodes": Array [],
  "extensions": undefined,
}
`;
